src/hdmf/data_utils.py

```diff
@@ -125,6 +127,272 @@ def maxshape(self):
+        )
+
+        if buffer_gb is None and buffer_shape is None:
+            buffer_gb = 1.0
+        if chunk_mb is None and chunk_shape is None:
+            chunk_mb = 1.0
+        assert (buffer_gb is not None) != (
+            buffer_shape is not None
+        ), "Only one of 'buffer_gb' or 'buffer_shape' can be specified!"
+        assert (chunk_mb is not None) != (
+            chunk_shape is not None
+        ), "Only one of 'chunk_mb' or 'chunk_shape' can be specified!"
+
+        self._maxshape = self._get_maxshape()
+        self._dtype = self._get_dtype()
+        if chunk_shape is None:
+            self.chunk_shape = self._get_default_chunk_shape(chunk_mb=chunk_mb)
+        else:
+            self.chunk_shape = chunk_shape
+        if buffer_shape is None:
+            self.buffer_shape = self._get_default_buffer_shape(buffer_gb=buffer_gb)
+        else:
+            self.buffer_shape = buffer_shape
+            buffer_gb = np.prod(self.buffer_shape) * np.dtype(self._dtype).itemsize / 1e9
+
+        array_chunk_shape = np.array(self.chunk_shape)
+        array_buffer_shape = np.array(self.buffer_shape)
+        array_maxshape = np.array(self.maxshape)
+        assert all(array_buffer_shape > 0), f"Some dimensions of buffer_shape ({self.buffer_shape}) are less than zero!"
+        assert all(
+            array_buffer_shape <= array_maxshape
+        ), f"Some dimensions of buffer_shape ({self.buffer_shape}) exceed the data dimensions ({self.maxshape})!"
+        assert all(
+            array_chunk_shape <= array_buffer_shape
+        ), f"Some dimensions of chunk_shape ({self.chunk_shape}) exceed the manual buffer shape ({self.buffer_shape})!"
+        assert all((array_buffer_shape % array_chunk_shape == 0)[array_buffer_shape != array_maxshape]), (
+            f"Some dimensions of chunk_shape ({self.chunk_shape}) do not "
+            f"evenly divide the buffer shape ({self.buffer_shape})!"
+        )
+
+        self.num_buffers = np.prod(np.ceil(array_maxshape / array_buffer_shape))
+        self.buffer_selection_generator = (
+            tuple([slice(lower_bound, upper_bound) for lower_bound, upper_bound in zip(lower_bounds, upper_bounds)])
+            for lower_bounds, upper_bounds in zip(
+                product(
+                    *[
+                        range(0, max_shape_axis, buffer_shape_axis)
+                        for max_shape_axis, buffer_shape_axis in zip(self.maxshape, self.buffer_shape)
+                    ]
+                ),
+                product(
+                    *[
+                        chain(range(buffer_shape_axis, max_shape_axis, buffer_shape_axis), [max_shape_axis])
+                        for max_shape_axis, buffer_shape_axis in zip(self.maxshape, self.buffer_shape)
+                    ]
+                ),
+            )
+        )
+
+        if self.display_progress:
```

Comment:
The progress bar option is nice but should be tested at least to make sure the code runs without error. This would also help us ensure high test code coverage.

I recommend adding tqdm to `requirements-dev.txt` and adding a test that sets `display_progress=True`. Also add a test on whether "total" is present in `progress_bar_options`.

Reply:
I cannot seem to figure out how to run the GitHub Action that runs the coverage tool on this PR from your fork. From what I can tell, it should have been run, or present me with the option to approve the workflow.

Reply:
I think I got the GH Action to run such that the coverage report from your fork is uploaded to the HDMF codecov page. You should be able to view the coverage report here: https://app.codecov.io/gh/hdmf-dev/hdmf/commit/e229b771bec984fda7affab75090f5c0d82dae43
and browse which lines of `data_utils.py` have changed in code coverage.

You can also run
```
python -m coverage run test.py -u
coverage report -m  # shell output
coverage html -d tests/coverage/htmlcov  # html output
```
locally and view a local version of this report.

Reply:
Thanks for the info!